src/linux/native/jgroup/com/alibaba/tenant/jgroup.c

```diff
@@ -0,0 +1,402 @@
+    }
+
+    // cpuacct.*
+    if (IS_CONTROLLER_ENABLED(CG_CPUACCT)) {
+        fid = (*env)->GetStaticFieldID(env, clazz, "IS_CPUACCT_ENABLED", "Z");
+        (*env)->SetStaticBooleanField(env, clazz, fid, JNI_TRUE);
+        JGROUP_INFO("%s mountpoint=%s", cg_controller_names[CG_CPUACCT], cg_mount_points[CG_CPUACCT]);
+    }
+
+    return res;
+}
+
+// invoked once in static initializer
+JNIEXPORT jint JNICALL
+Java_com_alibaba_tenant_NativeDispatcher_init0(JNIEnv* env, jclass clazz) {
+    jint res = 0;
+
+    // Detect and initialize native configuration
+    if (0 != (res = detect_cgroup_enabled())) {
+        JGROUP_MSG("Failed to detect cgroup status!");
+        return res;
+    }
+
+    if (0 != (res = detect_cgroup_mount_points())) {
+        JGROUP_MSG("Failed to detect cgroup mountpoints!");
+        return res;
+    }
+
+    // detect debugging options from environment variables
+    char* true_str = "TRUE"; // will be compared in uppercase
+    char* env_val  = getenv(JGROUP_DEBUG_ENV_KEY);
+    if (env_val != NULL && strlen(true_str) == strlen(env_val)) {
+        size_t idx = 0;
+        jgroup_debug = JNI_TRUE;
+        for (; idx < strlen(true_str); ++idx) {
+            if (toupper(env_val[idx]) != true_str[idx]) {
+                jgroup_debug = JNI_FALSE;
+                break;
+            }
+        }
+    }
+
+    JGROUP_INFO("jgroup debugging mode = %s", jgroup_debug == JNI_TRUE ? "true" : "false");
+
+    // initialize cgroup
+    if (0 != (res = init_Java_globals(env, clazz))) {
+        JGROUP_MSG("Failed to init Java globals!");
+    } else {
+        JGROUP_MSG("cgroup initialized successfully");
+    }
+
+    return res;
+}
+
+//
+// Create the cgroup dir, and initialize some mandatory values
+// @param group_path   relative path of target CGROUP
+JNIEXPORT jint JNICALL
+Java_com_alibaba_tenant_NativeDispatcher_createCgroup(JNIEnv* env,
+                                                       jobject ignored,
+                                                       jstring group_path) {
+    int res = 0;
+    int cg_index = 0;
+    struct stat st;
+    if (group_path == NULL) {
+        JGROUP_MSG("null group path to create");
+        res = -1;
+    } else {
+        const char* path_str = (*env)->GetStringUTFChars(env, group_path, NULL);
+        if (path_str != NULL || strlen(path_str) > 0) {
+            char path_buf[MAX_PATH_LEN];
+            for (cg_index = 0; cg_index < CG_MAX_SUPPORTED; ++cg_index) {
+                snprintf(path_buf, MAX_PATH_LEN, "%s/%s", cg_mount_points[cg_index], path_str);
+
+                // if exists, just reuse
+                if (stat(path_buf, &st) == 0 && S_ISDIR(st.st_mode)) {
+                    JGROUP_INFO("Reuse existing directory %s", path_buf);
+                    continue;
+                }
+
+                res = mkdir(path_buf,
+                            S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);  // modes are copied from libcg
```

Comment:
should it be placed at the same line with 'mkdir' call.

Reply:
done